gk/bpf.c

```diff
@@ -423,6 +426,59 @@ gk_init_bpf_cookie(const struct gk_config *gk_conf, uint8_t program_index,
 	return 0;
 }
 
+static int
+parse_packet_further(struct ipacket *packet, struct gk_bpf_pkt_ctx *ctx)
+{
+	struct rte_mbuf *pkt = packet->pkt;
+	uint16_t parsed_len = pkt_in_l2_hdr_len(pkt);
+
+	pkt->l2_len = parsed_len;
+	ctx->l3_proto = packet->flow.proto;
+
+	/*
+	 * extract_packet_info() guarantees that the L2 header and
+	 * the L3 headers without extensions are in the packet.
+	 */
+
+	switch (packet->flow.proto) {
+	case RTE_ETHER_TYPE_IPV4: {
+		struct rte_ipv4_hdr *ipv4_hdr = rte_pktmbuf_mtod_offset(pkt,
+			struct rte_ipv4_hdr *, parsed_len);
+		pkt->l3_len = ipv4_hdr_len(ipv4_hdr);
+		parsed_len += pkt->l3_len;
+		ctx->fragmented = rte_ipv4_frag_pkt_is_fragmented(ipv4_hdr);
+		ctx->l4_proto = ipv4_hdr->next_proto_id;
+		break;
+	}
+
+	case RTE_ETHER_TYPE_IPV6: {
+		struct rte_ipv6_hdr *ipv6_hdr = rte_pktmbuf_mtod_offset(pkt,
+			struct rte_ipv6_hdr *, parsed_len);
+		int l3_len = ipv6_skip_exthdr(ipv6_hdr,
+			pkt->data_len - parsed_len, &ctx->l4_proto);
+		if (l3_len < 0) {
+			GK_LOG(NOTICE, "%s: Failed to parse IPv6 extension headers\n",
+				__func__);
+			return -1;
+		}
+		pkt->l3_len = l3_len;
+		parsed_len += l3_len;
+		ctx->fragmented = rte_ipv6_frag_get_ipv6_fragment_header(
+			ipv6_hdr) != NULL;
+		break;
+	}
+
+	default:
+		GK_LOG(ERR, "%s: Unknown L3 header %hu\n",
+			__func__, packet->flow.proto);
+		return -1;
+	}
+
+	pkt->l4_len = RTE_MIN(pkt->data_len - parsed_len,
+		((1 < RTE_MBUF_L4_LEN_BITS) - 1));
```

Comment:
I don't quite catch what's happening here. Is this supposed to be a bit shift (`<<`)? If so, why by this constant?

Reply:
Yes. It was supposed to be a bit shift. Good catch.

Field `l4_len` is only `RTE_MBUF_L4_LEN_BITS` long, so `((1 < RTE_MBUF_L4_LEN_BITS) - 1)` is the maximum value the field can hold. `RTE_MIN()` is used because `pkt->data_len - parsed_len` may be bigger than what `l4_len` can hold.